File: windows/operations.py
"""Account status and local recovery. Diagnostic records contain only fixed messages."""
import ui_language
import copy, datetime as dt, hashlib, json, re, socket, threading, time, uuid
from pathlib import Path

LABELS = {'queued':ui_language.t('Venter'), 'retry':ui_language.t('Venter på nytt forsøk'), 'reserved':ui_language.t('Venter på sending'),
          'sent':ui_language.t('Sendt til LoTW – behandling må kontrolleres'), 'manual-batch':ui_language.t('Krever historisk import hos Club Log'),
          'starting':ui_language.t('Sendt – resultat må kontrolleres'), 'sending':ui_language.t('Sender – venter på kvittering'),
          'ack':ui_language.t('Leveringsresultat venter på Cloud'),
          'delivered':ui_language.t('Bekreftet lagret'), 'uncertain':ui_language.t('Ukjent resultat – kontroller mottaker'),
          'rejected':ui_language.t('Avvist – kontroller innstillinger'), 'changed':ui_language.t('Endret etter levering')}

class History:
    def __init__(self, folder):
        self.path=Path(folder)/'error-history.json';self.lock=threading.RLock()
    def read(self):
        try:return json.loads(self.path.read_text(encoding='utf-8'))[-200:]
        except (OSError,ValueError):return []
    def record(self, area, error):
        from cloud_sync import atomic
        text=str(error).lower()
        if any(x in text for x in ('sertifikat','certificate','ssl','https')):code='tls';message=ui_language.t('Sikker forbindelse kunne ikke bekreftes.')
        elif any(x in text for x in ('passord','logg inn','innlogging','password','sign in','log in','login','401','403')):code='konto';message=ui_language.t('Innlogging eller tilgang må kontrolleres.')
        elif any(x in text for x in ('nett','network','server','tilkob','connection','timeout','timed out')):code='nett';message=ui_language.t('Serverforbindelsen feilet. Kontroller nett og serverstatus.')
        elif isinstance(error,OSError):code='fil';message=ui_language.t('En lokal fil kunne ikke leses eller lagres. Kontroller lagringsplass og filtilgang.')
        elif any(x in text for x in ('konflikt','duplikat','endret','conflict','duplicate','changed')):code='konflikt';message=ui_language.t('Ulike opplysninger trenger kontroll før de kan behandles.')
        else:code='operasjon';message=ui_language.t('Handlingen ble ikke fullført. Kontroller statusvisningen og prøv igjen.')
        area=ui_language.canonical(area);message=ui_language.canonical(message)
        area=area if area in ('Cloud','App','Radio','Gjenoppretting','Loggtjenester') else 'App'
        with self.lock:
            items=self.read();now=time.time()
            if items and items[-1]['area']==area and items[-1]['code']==code and now-items[-1]['last']<300:
                items[-1].update(last=now,count=items[-1]['count']+1)
            else:items.append({'time':now,'last':now,'area':area,'code':code,'message':message,'count':1})
            atomic(self.path,items[-200:])

class Operations:
    def __init__(self,sync):
        self.sync=sync;self.history=History(sync.folder.parent)
        path=sync.folder/'device.json'
        from cloud_sync import atomic
        if not path.exists():atomic(path,{'id':str(uuid.uuid4())})
        self.device=json.loads(path.read_text(encoding='utf-8'))['id']
    def namespace(self):
        c=self.sync.config
        if not c.get('user'):raise ValueError(ui_language.t('Logg inn først.'))
        return hashlib.sha256((c['url']+'|'+c['user']).encode()).hexdigest()
    def local(self):
        from qso_sync import canonical
        return {k:canonical(v) for k,v in self.sync.snapshot().items() if k.startswith('qso/') and v is not None}
    def baseline(self):
        path=self.sync.folder/(self.namespace()+'.json')
        return self.sync.read_private(path) if path.exists() else {}
    def client(self):
        stamp=self.sync.config.get('last_sync_epoch',0)
        baseline=self.baseline() if self.sync.config.get('user') else {}
        pending=sum(baseline.get(k,{}).get('value')!=v for k,v in self.local().items())
        return {'device':self.device,'name':socket.gethostname()[:64],'last_sync':stamp,'pending':pending}
    def overview(self):
        s=self.sync
        with s.lock:
            local=self.local();baseline=self.baseline();contacts=[]
            for key,q in local.items():
                synced=baseline.get(key,{}).get('value')==q
                contacts.append({'key':key,'call':q.get('CALL',''),'date':q.get('QSO_DATE',''),'time':q.get('TIME_ON',''),
                                 'cloud':ui_language.t('I Cloud') if synced else ui_language.t('Venter på Cloud')})
            pending=sum(q['cloud']!=ui_language.t('I Cloud') for q in contacts);s.status['pending_contacts']=pending
            try:devices=s.request('/v1/account/devices',{});device_error=''
            except Exception as e:
                devices={'items':[]};device_error=ui_language.t('Enhetsoversikten kunne ikke hentes. Krever Cloud-server 0.6.2 eller nyere.')
            try:s.deliveries();delivery_error=''
            except Exception:delivery_error=ui_language.t('Leveringsstatus kunne ikke oppdateres. Sist hentede status vises.')
            deliveries=[{k:item.get(k) for k in ('qso','call','date','time','service','profile','state','updated','message')} for item in s.status.get('deliveries',[])]
            from log_delivery import Outbox
            by_id={(d['qso'],d['service'],d['profile']):d for d in deliveries}
            for d in deliveries:
                d['can_resolve']=d['state'] in ('uncertain','rejected')
                d['expected']=d['updated']
            for d in Outbox(s).rows():
                ident=(d['qso'],d['service'],d['profile'])
                remote=by_id.get(ident,{})
                if remote.get('state') not in ('delivered','sent'):
                    d.update(can_resolve=bool(remote.get('can_resolve')),expected=remote.get('expected'))
                    by_id[ident]=d
            deliveries=list(by_id.values())
            for item in deliveries:
                item['label']=ui_language.t(LABELS.get(item['state'],'Ingen bekreftelse'))
                if isinstance(item.get('message'),str):item['message']=ui_language.t(item['message'])
            return {'contacts':sorted(contacts,key=lambda x:(x['date'],x['time']),reverse=True),'total':len(local),'pending':pending,
                    'last_sync':s.config.get('last_sync_epoch',0),'devices':devices,'device_error':device_error,
                    'deliveries':deliveries,'delivery_error':delivery_error,'conflicts':len(s.view().get('conflicts',[])),
                    'errors':[{**item,'area':ui_language.t(item['area']),'message':ui_language.t(item['message'])} for item in self.history.read()]}
    def backup_items(self):
        folder=self.sync.folder/'backups'/self.namespace();items=[]
        for p in sorted(folder.glob('*.json'),reverse=True):
            try:
                data=self.sync.read_private(p);count=sum(k.startswith('qso/') and v is not None for k,v in data['data'].items())
                items.append({'id':p.name,'time':p.stat().st_mtime,'contacts':count})
            except Exception:items.append({'id':p.name,'time':p.stat().st_mtime,'contacts':None,'error':ui_language.t('Sikkerhetskopien kunne ikke leses.')})
        return {'items':items}
    def backup_data(self,name):
        if not isinstance(name,str) or not re.fullmatch(r'[0-9TZ.]+\.json',name):raise ValueError(ui_language.t('Ugyldig sikkerhetskopi.'))
        folder=(self.sync.folder/'backups'/self.namespace()).resolve();path=(folder/name).resolve()
        if path.parent!=folder or not path.is_file():raise ValueError(ui_language.t('Sikkerhetskopien finnes ikke for denne kontoen.'))
        source=self.sync.read_private(path)
        if source.get('format')!=1 or not isinstance(source.get('data'),dict):raise ValueError(ui_language.t('Sikkerhetskopien er ugyldig.'))
        data={k:v for k,v in source['data'].items() if k.startswith('qso/') and v is not None}
        from shared_log import identity
        # Identity and field validation are also performed by the application's apply callback.
        for key,q in data.items():
            if not re.fullmatch('qso/[0-9a-f]{64}',key) or not isinstance(q,dict):raise ValueError(ui_language.t('Ugyldig kontakt i sikkerhetskopien.'))
            if key!='qso/'+hashlib.sha256(json.dumps(identity(q)).encode()).hexdigest():raise ValueError(ui_language.t('Kontaktens identitet stemmer ikke i sikkerhetskopien.'))
            try:dt.datetime.strptime(q['QSO_DATE']+q['TIME_ON'],'%Y%m%d%H%M%S')
            except (ValueError,KeyError):raise ValueError(ui_language.t('Ugyldig dato eller tid i sikkerhetskopien.')) from None
        return data,path.stat().st_mtime
    def preview(self,name):
        with self.sync.lock:
            data,when=self.backup_data(name);local=self.local()
            return {'id':name,'time':when,'contacts':len(data),'current':len(local),'add':len(set(data)-set(local)),
                    'different':sum(k in local and local[k]!=v for k,v in data.items()),'preserved':len(local)}
    def restore(self,name):
        from cloud_sync import atomic
        with self.sync.lock:
            data,_=self.backup_data(name);captured=self.sync.snapshot();incoming={k:v for k,v in data.items() if captured.get(k) is None}
            stamp=dt.datetime.now(dt.timezone.utc).strftime('%Y%m%dT%H%M%S.%fZ')
            self.sync.save_private(self.sync.folder/'backups'/self.namespace()/(stamp+'.json'),{'format':1,'data':captured})
            applied=self.sync.apply(incoming,captured)
            return {'added':len(applied),'preserved':sum(k.startswith('qso/') and v is not None for k,v in captured.items()),
                    'skipped':len(incoming)-len(applied),'message':ui_language.t('Manglende kontakter er hentet. Nåværende kontakter og innstillinger er beholdt.')}
    def diagnostics(self):
        s=self.sync
        return {'product':'HamNavigator MY SHACK','version':(Path(__file__).parent/'VERSION').read_text().strip(),
                'platform':'Windows','connected':bool(s.config.get('token')),'authorized':s.authorized,
                'last_sync':s.config.get('last_sync_epoch',0),'errors':self.history.read()}

File: windows/operations.js
'use strict';
(() => {
  pageNames.operations='Status og sikkerhetskopi';
  const nav=document.createElement('button');nav.dataset.page='operations';nav.textContent='◉ Status og sikkerhetskopi';document.querySelector('nav').append(nav);
  const section=document.createElement('section');section.id='page-operations';section.className='page';
  section.innerHTML=`<div class="page-heading"><div><p class="eyebrow">DINE ENHETER OG KONTAKTER</p><h1>Status og sikkerhetskopi</h1><p>Se hva som er lagret lokalt, synkronisert og levert.</p></div></div>
  <p id="ops-message" role="status"></p><div class="actions"><button id="ops-refresh" class="primary">Oppdater status</button><button id="ops-copy" class="secondary">Kopier feilsøkingsinformasjon</button></div>
  <article class="card"><h2>Synkronisering</h2><p id="ops-summary"></p><p id="ops-device-error"></p><div id="ops-devices"></div></article>
  <article class="card"><h2>Kontakter og loggtjenester</h2><label>Søk etter kallesignal eller dato<input id="ops-search" placeholder="Kallesignal eller ÅÅÅÅMMDD"></label><p>«I Cloud» bekrefter synkronisering. «Bekreftet lagret» betyr at loggtjenesten har kvittert eller at mottakerens logg er kontrollert manuelt. Det er ikke en QSL-bekreftelse fra motstasjonen.</p><p id="ops-delivery-error"></p><div id="ops-contacts"></div><button id="ops-more" class="secondary">Vis flere</button></article>
  <article class="card"><h2>Feilhistorikk</h2><p>Feil beholdes etter omstart. Like gjentakelser samles. Kopien inneholder ingen kontoopplysninger, passord eller API-nøkler.</p><div id="ops-errors"></div></article>
  <article class="card"><h2>Lokale sikkerhetskopier</h2><p>Kontroller dato og antall før du henter manglende kontakter. Kontakter som allerede finnes, nyere kontakter og dagens innstillinger beholdes. Det tas en kopi før gjenoppretting.</p><button id="ops-backups" class="secondary">Vis sikkerhetskopier på denne PC-en</button><div id="ops-backup-list"></div></article>`;
  document.querySelector('main').insertBefore(section,document.querySelector('main footer'));
  const $=s=>section.querySelector(s),date=n=>n?new Date(n*1000).toLocaleString('nb-NO'):'Ikke registrert';
  let busy=false,data=null,limit=30;
  const text=(parent,value,tag='p')=>{const el=document.createElement(tag);el.textContent=value;parent.append(el);return el;};
  async function call(action,body={}){const response=await fetch('/api/cloud/'+action,{method:'POST',headers:{'Content-Type':'application/json','X-Radio-Token':document.querySelector('meta[name="radio-token"]').content},body:JSON.stringify(body)});const result=await response.json();if(!response.ok)throw Error(result.error||'Handlingen feilet.');return result;}
  async function run(action){if(busy)return;busy=true;$('#ops-message').textContent='Arbeider …';try{await action();$('#ops-message').textContent='Status oppdatert.';}catch(e){$('#ops-message').textContent=e.message;}finally{busy=false;}}
  function contacts(){
    if(!data)return;const list=$('#ops-contacts'),expanded=new Set([...list.querySelectorAll('details[open]')].map(x=>x.dataset.key));list.replaceChildren();const query=$('#ops-search').value.trim().toUpperCase();
    const all=data.contacts.filter(q=>[q.call,q.date,q.time].join(' ').toUpperCase().includes(query));
    for(const q of all.slice(0,limit)){const box=document.createElement('details');box.dataset.key=q.key;box.open=expanded.has(q.key);list.append(box);text(box,`${q.call} · ${q.date} ${q.time} UTC · ${q.cloud}`,'summary');
      const deliveries=data.deliveries.filter(d=>d.qso===q.key);
      if(!deliveries.length)text(box,'Ingen leveringskvittering registrert i Cloud. Dette beviser ikke at kontakten mangler hos en loggtjeneste.');
      for(const d of deliveries){text(box,`${d.service} · ${d.label} · ${date(d.updated)}${d.message?' · '+d.message:''}`);
        if(d.can_resolve){const b=text(box,'Avklar levering','button');b.className='secondary';b.onclick=()=>{
          const dialog=document.createElement('dialog');text(dialog,`${q.call} · ${q.date} ${q.time} UTC · ${d.service}`,'h2');text(dialog,'Kontroller denne kontakten i mottakerens logg før du velger. Ved ukjent resultat sendes den ikke automatisk på nytt.');
          for(const [state,label] of [['delivered','Kontakten finnes hos mottaker'],['retry','Kontakten finnes ikke – tillat nytt forsøk']]){const choice=text(dialog,label,'button');choice.onclick=()=>{dialog.close();dialog.remove();run(async()=>{await call('delivery-choice',{qso:d.qso,service:d.service,profile:d.profile,expected:d.expected,checked:true,state});data=await call('overview');render();});};}
          text(dialog,'Avbryt','button').onclick=()=>{dialog.close();dialog.remove();};document.body.append(dialog);dialog.showModal();};}
      }
    }$('#ops-more').hidden=all.length<=limit;if(!all.length)text(list,'Ingen kontakter passer søket.');
  }
  function render(){
    $('#ops-summary').textContent=`${data.total} kontakter · ${data.pending} venter på Cloud · ${data.conflicts} konflikter. Sist synkronisert: ${date(data.last_sync)}.`;
    $('#ops-device-error').textContent=data.device_error;$('#ops-delivery-error').textContent=data.delivery_error;
    const devices=$('#ops-devices');devices.replaceChildren();for(const d of data.devices.items||[])text(devices,`${d.name||d.platform||'Ukjent enhet'}${d.current?' (denne enheten)':''} · ${d.platform} ${d.version} · ${d.active?'Nylig aktiv':'Ikke nylig aktiv'} · sist sett ${date(d.seen)} · sist synkronisert ${date(d.last_sync)} · ${d.pending} venter`);
    const errors=$('#ops-errors');errors.replaceChildren();for(const e of [...data.errors].reverse())text(errors,`${date(e.last)} · ${e.area} · ${e.message} (${e.count} gang${e.count===1?'':'er'})`);if(!data.errors.length)text(errors,'Ingen feil registrert.');contacts();
  }
  nav.onclick=()=>{showPage('operations');run(async()=>{data=await call('overview');render();});};
  $('#ops-refresh').onclick=()=>run(async()=>{data=await call('overview');render();});
  $('#ops-search').oninput=()=>{limit=30;contacts();};$('#ops-more').onclick=()=>{limit+=30;contacts();};
  $('#ops-copy').onclick=()=>run(async()=>{const d=await call('diagnostics');await navigator.clipboard.writeText(JSON.stringify(d,null,2));toast('Feilsøkingsinformasjon kopiert.');});
  $('#ops-backups').onclick=()=>run(async()=>{const result=await call('backups'),list=$('#ops-backup-list');list.replaceChildren();
    for(const item of result.items){const row=text(list,`${date(item.time)} · ${item.contacts===null?'Kan ikke leses':item.contacts+' kontakter'}`);if(item.error)continue;
      const button=text(row,'Kontroller og gjenopprett','button');button.className='secondary';button.onclick=()=>run(async()=>{const p=await call('backup-preview',{id:item.id});
        if(!confirm(`Sikkerhetskopi fra ${date(p.time)}: ${p.contacts} kontakter.\n${p.add} manglende kontakter legges til.\nAlle ${p.current} nåværende kontakter beholdes. ${p.different} kontakter med ulike opplysninger beholder dagens versjon.\n\nHente de manglende kontaktene?`))return;
        const restored=await call('backup-restore',{id:item.id,confirm:true});await refreshState();toast(`${restored.added} kontakter hentet. ${restored.preserved} beholdt.`);data=await call('overview');render();});
    }if(!result.items.length)text(list,'Ingen Cloud-sikkerhetskopier på denne PC-en for den innloggede kontoen.');});
  setInterval(()=>{if(section.classList.contains('active')&&!busy&&!document.hidden&&!document.querySelector('dialog[open]'))run(async()=>{data=await call('overview');render();});},15000);
})();
